### ISSUE TITLE
Restrict use of walrus operator

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ruff#13741]
i see code written like this:
```py
a = (b := c)
```
who asked? (noone) it should be `a = b = c`

i see code written like this:
```py
foo(bar := baz()).bat()
bar.qux()
```

again, nobody has asked for this and it should just simply be an assignment on the line before. perhaps this could be argued to be stylistic, but that would only be true if you were arguing that bad style should be preferred

i'm not asking for walruses to be hunted to extinction, just the blatantly abusive cases be taken to account. occasionally, i see this:

```py
if match1 := pattern1.match(data):
    result = match1.group(1)
elif match2 := pattern2.match(data):
    result = match2.group(2)
else:
    result = None
```
(i saw it in [the pep](https://peps.python.org/pep-0572/))

this is okay, i don't have any issue with this and it can stay

# tl;dr

i'm not 100 certain what stays and what goes, but i think it would be roughly:
- if the walrus could be a normal assignment on the previous line, then it shouldn't be a walrus:
  ```py
  (a := 1)  # not ok
  
  if bool(): ...
  elif a := bool():  # ok
      print(a)
  ```
- except, for `if` conditions:
    ```py
    if foo := bar():
        print(foo)
    ```
  but even then, that looks pretty stupid to me, but on the border of being acceptable

**Comment by MichaReiser** [comment]
I'll merge this into https://github.com/astral-sh/ruff/issues/3464 and mention that an alternative is to restricting walrus usage only in specific places.
